### PR TITLE
fix(ci): exempt a change confined to a cfg(test) module from the release gate (#461)

### LINKED ISSUE #461: Release gate treats a test-only change inside a shipped file as shipped
The `version bump + release notes` job states its own intent in the comment
above it: "Docs / CI / test-only changes ship nothing and are exempt." It then
encodes "test-only" as a path list:

```
shipped="$(printf '%s\n' "$changed" \
  | grep -E '^(src/|assets/|build\.rs$|Cargo\.toml$|Cargo\.lock$)' \
  | grep -vE '^(src/app/tests\.rs|tests/)' || true)"
```

A change confined to a `#[cfg(test)] mod tests` block inside a shipped file is
test-only in fact, and the release binary it produces is byte-identical, but it
does not match either exempt path. The gate then demands a version bump and a
`src/release_notes/<version>.md` for a change that the welcome panel has
nothing true to say about: the notes file is a user-facing highlight list, so
the only way to satisfy the gate is to describe a change that does not exist in
the binary.

Live instance: #459 changes one assertion inside `mod tests` in
`src/markdown.rs` and nothing else. Every other check is green and the job
fails with:

```
Shipped files changed:
  src/markdown.rs
This PR changes shipped files but the version is still 0.1.857.
This PR changes shipped files but does not write src/release_notes/0.1.857.md.
```

Croft keeps most unit tests beside the code they cover, so this is the ordinary
shape of a test-only fix here, not an edge case. `src/app/tests.rs` is exempt
only because it happens to be a whole file.

## Suggested shape

Filter the shipped list once more: a `.rs` file whose diff touches nothing
outside a `#[cfg(test)]` module ships nothing. Everything else keeps the
current behaviour, and anything the parse is unsure about counts as shipped, so
the failure direction stays "asks for a bump it did not need" rather than
"waives one it needed".

Worth pinning with tests either way, since a gate that cannot fail and a gate
that waives too much look identical from the outside.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #461.

The `version bump + release notes` job says in its own comment that test-only
changes ship nothing and are exempt, and encodes that as a path list. A whole
test file (`src/app/tests.rs`, `tests/`) is therefore exempt, while a diff
inside a `#[cfg(test)] mod tests` in a shipped file is not. croft keeps most
unit tests beside the code they cover, so that is the ordinary shape of a
test-only fix here: the release binary is byte-identical, and the only way past
the gate is to bump a version and write user-facing release notes for a change
the binary does not contain.

`scripts/ships_nothing.py` answers, per file, whether every line the diff
touches lies inside a `#[cfg(test)]` module. The shipped list is filtered
through it before the version and notes are checked at all.

## The failure direction is deliberate

Waiving a bump that was needed puts two different binaries on one version,
which is the thing this job exists to prevent. Asking for a bump that was not
needed costs a version number. So every uncertainty resolves to "ships":

* a file that is not `.rs`;
* `#[cfg(test)]` on anything other than a module;
* a module whose opening brace is not on the `mod` line;
* a module still open at end of file.

Both sides of the diff are read, because a deletion leaves no line at the head
to inspect and a head-only check would waive the removal of shipped code as
readily as the removal of a test. Braces are counted through the doc gate's
literal-aware tracker, so a `format!("{}")` inside a test cannot close the
module early and leave the rest of the file looking like test code.

## Proved on the PR that is stuck on it

#459 changes one assertion inside `mod tests` in `src/markdown.rs` and nothing
else. Running the job's logic against that branch, with and without the filter:

```
$ gate_sim <merge-base> origin/fix/substring-or-assertion   # with the filter
src/markdown.rs: change confined to #[cfg(test)]; ships nothing.
No shipped files changed; no release required.
exit=0

$ gate_sim <merge-base> origin/fix/substring-or-assertion   # today's logic
Shipped files changed:
  src/markdown.rs
ERROR: version still 0.1.857
ERROR: no src/release_notes/0.1.857.md
exit=1
```

And the control that matters more, since a filter that waives everything would
also turn the first run green. The same branch with one non-test line added to
`src/markdown.rs`, filter available:

```
Shipped files changed:
  src/markdown.rs
ERROR: version still 0.1.857
exit=1
```

## Tests

13 tests in `scripts/tests/test_ships_nothing.py`, run by the existing
`python3 -m unittest discover -s scripts/tests` step. Four of them pin the
conservative direction rather than the exemption, and each was proved
load-bearing by removing the guard it covers and watching that test, and only
that test, fail:

| guard removed | test that failed |
| --- | --- |
| reading the base side of the diff | `test_deleting_shipped_code_ships` |
| the non-Rust file check | `test_a_file_that_is_not_rust_ships` |
| literal-aware brace counting | `test_a_brace_in_a_string_does_not_end_the_block_early` |
| requiring a `mod` after the attribute | `test_a_cfg_test_item_that_is_not_a_module_is_not_a_range` |

The deletion test had to be a pure deletion: an edited line leaves a
replacement at the head that a head-only check catches anyway, and the first
version of that test passed under the mutation for exactly that reason.

CI only: `scripts/`, `.github/` and `CONTRIBUTING.md` ship nothing, so there is
no version bump.

## Summary by CodeRabbit

* **Release Process**
  * Test-only Rust changes are now recognized as not affecting the shipped binary.
  * Release version and notes requirements are automatically waived for qualifying test-only changes.
  * Changes that cannot be confidently classified continue to be treated as shipped.
  * File handling in release checks is more reliable for paths containing spaces or special characters.
* **Tests**
  * Added comprehensive coverage for test-only detection, ambiguous changes, binary files, additions, deletions, and edge cases.

### CODE REVIEW

**Comment by vitali87** [comment]
## 🤖 Fallback review (round 1) — Claude Opus 5

Ran because CodeRabbit is rate-limited on this PR: its check reports `pass` annotated "Review rate limited", which is a green row with no review behind it. Verdict was **REQUEST_CHANGES**, one blocker and two majors. All are fixed in `47102f0`, each reproduced before being fixed.

**blocker — a diff whose only change is the `#[cfg(test)]` line was waived, and both directions change the binary.** The span starts at the attribute line on purpose, so the line that decides whether the module compiles sits inside the span it opens. Adding `#[cfg(test)]` above a module that ships today takes it out of the binary; removing it from a test module puts that module and everything in it in. Both were reproduced on throwaway repos and both returned "ships nothing".

A touched attribute line is now test-only only when every line of its module moved with it, which is exactly what a test module added or deleted in one piece looks like, and that is the case the exemption exists for. Starting the span at the `mod` line instead would have rejected a newly added test module, which is the main use case.

That fix exposed a second thing: the blank line above a new test module is an added line outside every span, so a whole added module shipped. Blank lines are excluded from the comparison first, since a line carrying no code cannot change the binary either way. The control for both is in the same run: `test_a_whole_test_module_added_at_once_still_ships_nothing`.

**major — an empty diff meant two different things.** `git diff -U0 <base> HEAD -- <path that matches nothing>` exits 0 with no output, and the script read that as "nothing changed" and waived. It now waives an empty diff only when the path exists at both ends. The job's `for f in $shipped` was the route to such a path: unquoted, so it word-split on spaces and glob-expanded `[`, `*` and `?`. It reads a line at a time now.

**major — the CONTRIBUTING sentence promised a conservatism the filter did not hold.** True as written once the blocker is fixed, so the sentence stays and the code now matches it.

**minor — `git()` tolerated a missing path on every call**, re-widening what `check_doc_ownership.git` narrows deliberately. Opt-in now, as there. Worth saying plainly: no test pins this one, because every input I could construct fails identically with and without it. It is defensive, and it is in because the sibling's reasoning applies here with more force, not because a test demanded it.

**minor — `classify` labels `///` as DOC, not SKIP**, so a documented test module was never recognised and lost the exemption silently. DOC joins the step-over, with a test.

### The three vacuous guards

All three the review named are now pinned, and each fails only its own test when its guard is removed:

| guard | test |
| --- | --- |
| a module still open at EOF is not a range | `test_a_module_left_open_at_end_of_file_is_not_a_range` |
| `mod tests;` with no body is not a range | `test_a_module_declaration_without_a_body_is_not_a_range` |
| the CLI's reject direction | `test_the_cli_reports_the_verdict_as_its_exit_status` (both directions in one run) |

Seven guards in total now carry a control, and one mutation I ran first was itself vacuous: cutting the blank-line filter on the removed side left the whole-module test green, because that test only adds lines. Cutting it on the added side fails it.

### Not taken here

Batching the filter into one process (the nit about 20-plus interpreter starts) is a real saving and is not in this change: it turns a per-file question into a protocol, and this PR is already the one deciding what "ships nothing" means.

### End to end

The job's own logic, run over the branch this exists for and over a real feature branch, with the new loop:

```
#459 (one assertion inside `mod tests` in src/markdown.rs)
  src/markdown.rs: change confined to #[cfg(test)]; ships nothing.
  No shipped files changed; no release required.

#457 (a feature branch)
  Shipped files changed:
    Cargo.lock
    Cargo.toml
    src/app/mod.rs
    src/release_notes/0.1.861.md
    src/widgets/command_palette.rs
```

**Comment by vitali87** [comment]
## 🤖 Fallback review (round 2) — Claude Opus 5

Verdict was **REQUEST_CHANGES**: one blocker, one major, three minors and three nits. All are fixed in `989582d`, and the blocker was mine from round 1.

**blocker — a diff of nothing but blank lines was waived with no span consulted.** Confirmed and reproduced before fixing. Excluding blank lines emptied both lists, and both `all(inside(...))` guards plus `_whole_span_moved` are then vacuously true, so the answer was "ships nothing" for a file with no test module in it at all. A blank line inside a multi-line string literal is not whitespace but content, and the 42 instances named in `src/pair.rs` and `src/highlight.rs` are exactly the shape where it changes what the binary does. A blank line now counts as whitespace only outside an open string literal, read from the brace tracker's own `open_string` state rather than re-derived.

The end-to-end control, on a `pair.rs`-shaped literal:

```
$ python3 scripts/ships_nothing.py main HEAD src/pair.rs   # one blank line removed from a raw string
verdict exit=1   (ships)
```

**major — the exemption could not fire for any module shaped like the ones in this repo.** Also confirmed, and measured rather than argued: the coverage check demanded every line of a span including blank ones, while the touched set had already had them removed. Counting the spans in `src/` at this head gives **152 of 152 containing at least one blank line**, so the main use case was unreachable. Blank lines are skipped in that check too, and the two whole-module tests now use a module with an interior blank line, which is the real shape.

**minor — a diff git describes without hunk headers.** `git diff -U0` on a file git treats as binary prints `Binary files ... differ` and no `@@`, which read as "nothing changed". The branch now depends on the diff text, so a described-but-unparsed change ships.

**minor — the brace tracker does not strip `/* */`.** Right: it handles strings, chars and `//`. A file containing a mid-line `/*` now yields no ranges at all, which fails to the safe side, with the finding's own fixture as the test.

**minor — the job's success line claimed a reason the status does not carry.** It said "confined to `#[cfg(test)]`" for files that reached exit 0 by other routes, including the one in the blocker. It says "ships nothing" now.

**minor — CONTRIBUTING's conservatism sentence.** True again with the blocker fixed, so it stays.

**nits** — `printf '  %s\n' $shipped` is quoted, the `mktemp` gets a `trap`, and `not (start in marked)` is `start not in marked`.

### Guards

Six now, each pinned by removing it and watching that test alone fail:

| guard | test |
| --- | --- |
| blank means whitespace only outside a literal | `test_a_blank_line_inside_a_string_literal_ships` |
| the removed side is filtered too | `test_deleting_a_whole_test_module_in_its_real_shape_ships_nothing` |
| span coverage skips blank lines | the two whole-module tests |
| a described-but-unparsed diff ships | `test_a_diff_git_reports_as_binary_ships` |
| a mid-line `/*` yields no ranges | `test_a_mid_line_block_comment_yields_no_ranges` |
| the attribute line alone is not test-only | the two attribute tests |

On the review's note that M2b was pinned by nothing: it is now, by the deletion test, which is the case where a blank line outside the span has to be forgiven for the module to be waived at all.

### End to end

```
#459  src/markdown.rs (one assertion inside `mod tests`)   -> ships nothing
#452  src/app/mod.rs  (a feature branch)                   -> ships
src/pair.rs, one blank line out of a raw string            -> ships
```

`allow_missing_path=False` remains unpinned, as disclosed last round: every input I can construct fails identically with and without it. It stays because the sibling gate's reasoning applies here with more force, not because a test demanded it.

**Comment by vitali87** [comment]
@coderabbitai full review

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -56,12 +57,39 @@ jobs:
...
+          # to prevent.
+          # Read the list a line at a time rather than word-splitting it: an
+          # unquoted `for f in $shipped` also expands globs, so a path
+          # holding `[`, `*` or `?` would reach the filter as a pathspec git
+          # cannot resolve, and a path holding a space would arrive as two.
+          kept="$(mktemp)"
+          trap 'rm -f "$kept"' EXIT
+          while IFS= read -r f; do
+            [ -n "$f" ] || continue
+            if python3 scripts/ships_nothing.py "$base" HEAD "$f"; then
```
[the remark this answers is not in the document]
Comment (vitali87) [reply .github/workflows/ci.yml:89]: Addressed in `49294e7`, though not the way the comment proposes, and the reason is the trigger.

`pull_request` runs the workflow file from the merge of the submitted branch, so a PR that wanted this gate neutered would edit the job rather than the helper it calls. Pinning `ships_nothing.py` to the base revision leaves the caller rewritable, so it would read as a protection that is not there.

What is worth removing is the CREDENTIAL. This job ran only inline shell before this PR, so it never took the posture the `doc comments stay with their function` job beside it documents for exactly this situation: it now carries `permissions: contents: read` and `persist-credentials: false`, so the script from the branch runs without a checkout token in the local git config to reach.

If the intent is stronger than that, the shape that actually holds is `pull_request_target` with a base checkout, which is a different security posture for the whole job and worth its own change rather than a half-measure here.
